File: Xadrez/Pecas.hs
module Main where

import Text.Printf

data Pecas = Pecas {
    tipo :: String,
    cor :: String,
    linha :: Int,
    coluna :: Int
}

inicializarPecas :: String -> String -> Int -> Int -> Pecas
inicializarPecas tipo cor linha coluna = Pecas tipo cor linha coluna

imprimirInformacoes :: Pecas -> IO ()
imprimirInformacoes peca = do
    imprimir "Tipo da peça: " (tipo peca)
    imprimir "Cor da peça: " (cor peca)
    printf "Posição: %d%d\n" (linha peca) (coluna peca)

imprimir :: String -> String -> IO ()
imprimir texto arg = putStrLn (texto ++ arg)

data Peao = Peao {
    super :: Pecas
}

inicializarPeao :: String -> String -> Int -> Int -> Peao
inicializarPeao tipo cor linha coluna = Peao (inicializarPecas tipo cor linha coluna)

verificarMovimento :: Peao -> Int -> Bool
verificarMovimento peao novaLinha
    | (linha (super peao) == 2 || linha (super peao) == 7) && (novaLinha == 4 || novaLinha == 5) = True
    | abs (novaLinha - linha (super peao)) == 1 = True
    | otherwise = False

movimentarPeao :: Peao -> Int -> IO ()
movimentarPeao peao novaLinha =
    if verificarMovimento peao novaLinha
        then do
            let novaPeca = (super peao) { linha = novaLinha }
            imprimir "Movimento Realizado"
            imprimirInformacoes novaPeca
        else imprimir "Movimento Inválido"

capturarPeca :: Peao -> Int -> Int -> String -> IO ()
capturarPeca peao colunaCapturar linhaCapturar pecaCapturada =
    if (abs (coluna (super peao) - colunaCapturar) == 1 || abs (linha (super peao) - linhaCapturar) == 1)
        then printf "A peça %s foi capturada com sucesso na linha %d coluna %d!\n" pecaCapturada linhaCapturar colunaCapturar
        else imprimir "Movimento inválido!"

File: Xadrez/Pecas.java
public class Pecas {
    String tipo;
    String cor;
    int linha;
    int coluna;

    public Pecas(String tipo, String cor, int linha, int coluna){
        this.tipo = tipo;
        this.cor = cor;
        this.linha = linha;
        this.coluna = coluna;
    }
    void imprimirInformacoes(){
        imprimir("Tipo da peça: " + this.tipo);
        imprimir("Cor da peça: " + this.cor);
        imprimir("Posição: " + this.linha + this.coluna);
    }
    void imprimir(String texto){
        System.out.println(texto);
    }
}
class Peao extends Pecas{
    public Peao(String tipo, String cor, int linha, int coluna){
        super(tipo, cor, linha, coluna);
    }
    void movimentarPeao(int novaLinha){
        if(verificarMovimento(novaLinha)){
            this.linha = novaLinha;
            imprimir("Movimento Realizado");
        } else{
            imprimir("Movimento Inválido");
        }
        
    }
    Boolean verificarMovimento(int novaLinha){
        if ((this.linha == 2 || this.linha == 7) && (novaLinha == 4 || novaLinha == 5)){
            return true;
        } else if (novaLinha - this.linha == 1 || this.linha - novaLinha == 1) {
            return true;
        } else{
            return false;
        }
    }
    void capturarPeca(int colunaCapturar, int linhaCapturar, String pecaCapturada){
        if(((colunaCapturar - this.coluna == 1) || (this.coluna - colunaCapturar == 1)) && ((linhaCapturar - this.linha == 1) || (this.linha - linhaCapturar == 1))){
            System.out.printf("A peça %s foi captura com sucesso na linha %d coluna %d!\n", pecaCapturada, linhaCapturar, colunaCapturar);
        } else{
            imprimir("Movimento inválido!");
        }
    }
}
